csharp/src/IceLocatorDiscovery/PluginI.cs

```diff
@@ -256,350 +173,224 @@ public void
-            {
-                if (!_pending)
+                Task t = await Task.WhenAny(completionSource.Task, Task.Delay(_timeout)).ConfigureAwait(false);
+                if (t == completionSource.Task)
                 {
-                    Debug.Assert(_pendingRequests.Count == 0);
-                    return; // Request failed
+                    return await completionSource.Task.ConfigureAwait(false);
                 }
 
-                if (_pendingRetryCount > 0)
+                lock (_mutex)
                 {
-                    --_pendingRetryCount;
-                    try
+                    if (_completionSource != null)
                     {
-                        if (_traceLevel > 1)
+                        // Locator lookup timeout and no more retries
+                        if (_traceLevel > 0)
                         {
-                            var s = new StringBuilder("retrying locator lookup:\nlookup = ");
+                            var s = new StringBuilder("locator lookup timed out:\nlookup = ");
                             s.Append(_lookup);
-                            s.Append("\nretry count = ").Append(_retryCount);
-                            if (_instanceName.Length == 0)
+                            if (_instanceName.Length > 0)
                             {
                                 s.Append("\ninstance name = ").Append(_instanceName);
                             }
                             _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
                         }
 
-                        foreach (KeyValuePair<ILookupPrx, ILookupReplyPrx?> l in _lookups)
-                        {
-                            _ = FindLocator(l.Key, l.Value!); // Sent multicast request
-                        }
-                        _timer.Schedule(this, _timeout);
-                        return;
-                    }
-                    catch (Exception)
-                    {
-                    }
-                    _pendingRetryCount = 0;
-                }
-
-                Debug.Assert(_pendingRetryCount == 0);
-                _pending = false;
-
-                if (_traceLevel > 0)
-                {
-                    var s = new StringBuilder("locator lookup timed out:\nlookup = ");
-                    s.Append(_lookup);
-                    if (_instanceName.Length == 0)
-                    {
-                        s.Append("\ninstance name = ").Append(_instanceName);
-                    }
-                    _lookup.Communicator.Logger.Trace("Lookup", s.ToString());
-                }
-
-                if (_pendingRequests.Count == 0)
-                {
-                    Monitor.Pulse(this);
+                        _nextRetry = Time.CurrentMonotonicTimeMillis() + _retryDelay;
+                        _completionSource.SetResult(_voidLocator);
+                        _completionSource = null;
+                        _findLocatorTask = null;
```

Comment:
This code in the `for() {}` loop looks wrong. We should only clear the task completion source outside the loop since it's created outside the loop.

Reply:
The task is clean in a 3 places now:

When we got a reply

https://github.com/zeroc-ice/ice/blob/c4a7ee2288ee1544aa32011d67769476129d1ab6/csharp/src/IceLocatorDiscovery/PluginI.cs#L228

When we exit the loop because all lookup invocations failed 

https://github.com/zeroc-ice/ice/blob/c4a7ee2288ee1544aa32011d67769476129d1ab6/csharp/src/IceLocatorDiscovery/PluginI.cs#L322 

Outside the loop when we consume all retries

https://github.com/zeroc-ice/ice/blob/c4a7ee2288ee1544aa32011d67769476129d1ab6/csharp/src/IceLocatorDiscovery/PluginI.cs#L354

Reply:
I see, I misplaced this when split the method.